### PR TITLE
feat(ticket#101):Add Server Action to Retrieve All Members from Members Table

### PR DESCRIPTION
**Pull request by nadirhussain786**
## Description:
This pull request introduces a new server action to fetch all members from the Members table, leveraging the specified schema. This enhancement improves the application's ability to manage user data effectively.

## Changes Made:

**Created a Server Action:** Implemented logic to retrieve all members based on the provided schema.
**Schema Definitions:**

- UserRole: Enum defining user roles (`member` , `shift_incharge`, `incharge`).
- member_shifts: Enum for shift types (`shift_a` , `shift_b` , `shift_c` , `shift_d` ).
- UserStatus: Enum for user status (`active`, `inactive`).
- User: Interface defining the structure of a user object with properties:
- name: string
- phone: string
- cnic: number
- shift: member_shifts
- role: UserRole
- address: string
- status: UserStatus

### CODE REVIEW

File: src/app/member/components/MemberList.tsx [hunk 1]
```diff
@@ -19,131 +19,14 @@ import {
...
-    status: "active",
-  },
-];
-
-export const MemberList = () => {
+interface MemberProps {
+  data: Member[];
+  error?: string;
+}
+export const MemberList = ({ data, error }: MemberProps) => {
```
Comment (muhammad-bilal-ali-khokhar) [inline src/app/member/components/MemberList.tsx:28]: @nadirhussain786 why are you passing `data` and `error` we have server action so can we call it directly in this component

File: src/app/member/components/MemberList.tsx [hunk 2]
```diff
@@ -199,12 +82,21 @@ export const MemberList = () => {
...
 
   const table = useReactTable({
-    data: members,
+    data,
     columns,
     getCoreRowModel: getCoreRowModel(),
   });
 
-  const groupedData = useGroupedData(members, "shift");
+  useEffect(() => {
```
Comment (muhammad-bilal-ali-khokhar) [inline src/app/member/components/MemberList.tsx:90]: @nadirhussain786 we have hook for toast , check that

File: src/app/member/loading.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            <TableHead>
+              <Skeleton className="h-4 w-24" />
+            </TableHead>
+            <TableHead className="text-right">
+              <Skeleton className="h-4 w-16 ml-auto" />
+            </TableHead>
+          </TableRow>
+        </TableHeader>
+        <TableBody>
+          {[...Array(5)].map((_, i) => (
```
Comment (muhammad-bilal-ali-khokhar) [inline src/app/member/loading.tsx:32]: @nadirhussain786  not good approch

File: src/app/member/components/MemberList.tsx [hunk 4]
```diff
@@ -252,6 +144,13 @@ export const MemberList = () => {
             ))}
           </React.Fragment>
```
Comment (muhammad-bilal-ali-khokhar) [inline src/app/member/components/MemberList.tsx:145]: @nadirhussain786  I am unable to see opening `React.Fragment` tag

File: src/app/member/page.tsx [hunk 5]
```diff
@@ -1,9 +1,13 @@
-import React from "react";
+import React, { Suspense } from "react";
```
Comment (muhammad-bilal-ali-khokhar) [inline src/app/member/page.tsx:1]: @nadirhussain786  remove unwanted imports

**Review by muhammad-bilal-ali-khokhar — COMMENTED** [review]
Please add attachements

**Review by muhammad-bilal-ali-khokhar — APPROVED** [review]

File: src/app/layout.tsx [hunk 6]
```diff
@@ -10,7 +10,11 @@ export const metadata: Metadata = {
   description: "Generated by create next app",
 };
 
-export default function RootLayout({
+export const dynamic = "force-dynamic";
```
Comment (Monfernape) [inline src/app/layout.tsx:13]: Why do we need these two?

File: src/app/member/actions/getMembers.test.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    const mockedGetMembers = getMembers as Mock;
+    const mockError = { error: "Failed to fetch", data: [] };
+
+    mockedGetMembers.mockResolvedValue(mockError);
+
+    const result = await mockedGetMembers();
+    console.log("Result from getMembers:", result);
+
+    expect(result.error).toBeDefined();
+    expect(result.data).toEqual([]);
```
Comment (Monfernape) [inline src/app/member/actions/getMembers.test.ts:35]: Nice

File: src/app/member/actions/getMembers.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
+"use server";
+
+import { getSupabaseClient } from "@/utils/supabase/supabaseClient";
+
+export const getMembers = async () => {
+  const supabase = await getSupabaseClient();
+  const { data, error } = await supabase.from("members").select("*");
```
Comment (Monfernape) [inline src/app/member/actions/getMembers.ts:7]: We should use the DBTable enum here.

File: src/app/member/components/MemberList.tsx [hunk 9]
```diff
@@ -199,12 +85,21 @@ export const MemberList = () => {
...
   const table = useReactTable({
-    data: members,
+    data,
     columns,
     getCoreRowModel: getCoreRowModel(),
   });
 
-  const groupedData = useGroupedData(members, "shift");
+  useEffect(() => {
+if (!success) {
```
Comment (Monfernape) [inline src/app/member/components/MemberList.tsx:94]: Why do we need this? Only members need to be passed from server component.

File: src/app/member/components/MemberList.tsx [hunk 10]
```diff
@@ -199,12 +85,21 @@ export const MemberList = () => {
...
+    toast({
+      title: "No Members Found",
```
Comment (Monfernape) [inline src/app/member/components/MemberList.tsx:96]: We should handle it more like UI state. We don't need to rely on API. When you're fetching data, the component won't render, instead loading.tsx would be in process. You can place two  conditions here:
- If data, show table.
- else show Standardpage. @muhammad-bilal-ali-khokhar let's create a task to create and incorporate standard page.

File: src/app/member/loading.tsx [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  Table,
+  TableBody,
+  TableCell,
+  TableHead,
+  TableHeader,
+  TableRow,
+} from "@/components/ui/table";
+import { Skeleton } from "@/components/ui/skeleton";
+
+export default function Loading() {
```
Comment (Monfernape) [inline src/app/member/loading.tsx:11]: @muhammad-bilal-ali-khokhar I think we can create one global loading state and let's not make it look like Table.

**Review by Monfernape — COMMENTED** [review]
Left some comments, would do another pass.

File: src/app/member/components/MemberList.tsx (hunk 10 above)
Comment (muhammad-bilal-ali-khokhar) [reply src/app/member/components/MemberList.tsx:96]: Sir @Monfernape  , I have created `ticket` for standard page

File: src/app/member/page.tsx [hunk 12]
```diff
@@ -1,9 +1,13 @@
...
-const Page = () => {
+import { getMembers } from "@/app/member/actions/getMembers";
+
+const Page = async () => {
+  const { data, success , message } = await getMembers();
+
   return (
     <div className="p-8">
-      <MemberList />
+      <MemberList data={data} success={success} message={message} />
```
Comment (Mubashar1009) [inline src/app/member/page.tsx:10]: can we handle error case also here

**Pull request merged**